### ISSUE TITLE
dart format hangs on Windows when reading from stdin in the presubmit

### ISSUE THREAD

**Issue by athomas** [dart-lang/sdk#46947]
The script below hangs on Windows but works fine on other platforms (change the hard coded depot_tools location in the script before running it):

```python
#!/usr/bin/env python3

import imp
import os
import subprocess
import sys

depot_tools = '/Users/user/src/depot_tools'
sys.path.append(depot_tools)
import scm

utils = imp.load_source('utils', os.path.join('tools', 'utils.py'))
dart = os.path.join(utils.CheckedInSdkPath(), 'bin', 'dart')

windows = utils.GuessOS() == 'win32'
if windows:
    dart += '.exe'

args = [
    dart,
    'format',
    '--set-exit-if-changed',
    '--output=none',
    '--summary=none',
]

print('Body builder from git')
contents = scm.GIT.Capture(['show', 'c9d954efe7c7a71d2bf2c2ca385855b82fbfb379:pkg/front_end/lib/src/fasta/kernel/body_builder.dart'],
                                strip_out=False)
process = subprocess.run(args, input=contents, text=True)
print(process.returncode)
```

For some reason, dart format hangs reports weird formatting errors and starts to print messages like this:
```
Attempt:11 waiting for isolate vm-isolate to check in
Attempt:11 waiting for isolate dartdev to check in
```

On other platforms, the script exits normally and `dart format` finds no formatting issues.

@bkonyi @munificent Any ideas about what could go wrong here? Perhaps line ending confusion?

/cc @johnniwinther

**Comment by athomas** [comment]
Note that simpler test cases like these work even on Windows:

```python
contents = """
void main()                     {}
"""
print('Format change')
process = subprocess.run(args, input=contents, text=True)
print(process.returncode)
 
contents = """
void main()                     {
"""
print('Syntax errror')
process = subprocess.run(args, input=contents, text=True)
print(process.returncode)
 
contents = """void main() {}
"""
print('No change')
process = subprocess.run(args, input=contents, text=True)
print(process.returncode)
```

**Comment by mraleph** [comment]
Is this a duplicate of https://github.com/dart-lang/sdk/issues/46600 (though that one was marked as fixed).

**Comment by athomas** [comment]
I don't think so. Before #46600, this would hang on macOS as well. Now it seems to only happen on Windows for specific inputs.

Also, a part of the fix was 44ffad59d2a5d54d7b68bd81e2875fa6d3886f9b which may have masked the issue. It removed the process.communicate causing the presubmit to no longer block on the `dart format` call. That broke the formatting check. This was fixed in d2bd43f43e6636036a3d250eadcd429c444b7540 which re-enabled the formatting checks.

**Comment by athomas** [comment]
The checked-in SDK is currently at `2.14.0-377.4.beta`, which has the dart_style fixes for #44600.

**Comment by johnniwinther** [comment]
Any update on this? I am forced to upload most of my CLs with `--bypass-hook` because of this, and who knows how many badly formatted files have been committed due to this.

**Comment by munificent** [comment]
Sorry for the delay. I was out on vacation.

> For some reason, dart format hangs reports weird formatting errors and starts to print messages like this:
> 
> ```
> Attempt:11 waiting for isolate vm-isolate to check in
> Attempt:11 waiting for isolate dartdev to check in
> ```

I've never seen anything like that before, sorry. :( I also don't have a Windows machine to try to repro this on.

For what it's worth, dart_style does have tests for reading from stdin and I believe those tests are run and pass on Windows in the SDK. However, those tests also weren't sufficient to catch #46600. I don't recall exactly why. I think maybe because the tests spawn `dart format` not using a shell, which affects the behavior?
